ocaml/testsuite/tests/typing-unique/overwriting.ml

```diff
@@ -661,6 +787,169 @@ Error: Overwrite may not change the tag to OptionA.
+|}]
+
+let mutating_tag_par r =
+  match r with
+  | { m = OptionA s } ->
+    (r.m <- OptionB s), overwrite_ r.m with OptionA s
+  | _ -> (), OptionB ""
+[%%expect{|
+Line 4, characters 44-51:
+4 |     (r.m <- OptionB s), overwrite_ r.m with OptionA s
+                                                ^^^^^^^
+Error: Overwrite may not change the tag to OptionA.
+Hint: The old tag of this allocation was changed through mutation.
+|}]
+
+let mutating_tag_par_parent r =
+  match r with
+  | { m = { x = OptionA s } } ->
+    (r.m <- { x = OptionB s }), overwrite_ r.m.x with OptionA s
+  | _ -> (), OptionB ""
+[%%expect{|
+Line 4, characters 54-61:
+4 |     (r.m <- { x = OptionB s }), overwrite_ r.m.x with OptionA s
+                                                          ^^^^^^^
+Error: Overwrite may not change the tag to OptionA.
+Hint: The old tag of this allocation was changed through mutation.
+|}]
+
+let mutating_tag_choice r =
+  match r with
+  | { m = OptionA s } ->
+    if true then (r.m <- OptionB s; r.m)
+            else overwrite_ r.m with OptionA s
+  | _ -> OptionB ""
+[%%expect{|
+Line 5, characters 37-44:
+5 |             else overwrite_ r.m with OptionA s
+                                         ^^^^^^^
+Error: Overwrite may not change the tag to OptionA.
+Hint: The old tag of this allocation was changed through mutation.
+|}]
+
+let mutating_tag_choice_parent r =
+  match r with
+  | { m = { x = OptionA s } } ->
+    if true then (r.m <- { x = OptionB s }; r.m.x)
+    else overwrite_ r.m.x with OptionA s
+  | _ -> OptionB ""
+[%%expect{|
+Line 5, characters 31-38:
+5 |     else overwrite_ r.m.x with OptionA s
+                                   ^^^^^^^
+Error: Overwrite may not change the tag to OptionA.
+Hint: The old tag of this allocation was changed through mutation.
+|}]
+
+let mutating_tag_choice_seq r =
+  match r with
+  | { m = OptionA s } ->
+    (if true then r.m <- OptionB s else ());
+    overwrite_ r.m with OptionA s
+  | _ -> OptionB ""
+[%%expect{|
+Line 5, characters 24-31:
+5 |     overwrite_ r.m with OptionA s
+                            ^^^^^^^
+Error: Overwrite may not change the tag to OptionA.
+Hint: The old tag of this allocation was changed through mutation.
+|}]
+
+let mutating_tag_choice_seq_parent r =
+  match r with
+  | { m = { x = OptionA s } } ->
+    (if true then r.m <- { x = OptionB s } else ());
+    overwrite_ r.m.x with OptionA s
+  | _ -> OptionB ""
+[%%expect{|
+Line 5, characters 26-33:
+5 |     overwrite_ r.m.x with OptionA s
+                              ^^^^^^^
+Error: Overwrite may not change the tag to OptionA.
+Hint: The old tag of this allocation was changed through mutation.
+|}]
+
+
+let mutating_tag_rematch r =
+  match r with
+  | { m = OptionA s } ->
+    r.m <- OptionB s;
+    begin match r with
+    | { m = OptionB s } ->
+      overwrite_ r.m with OptionB s
+    | _ -> OptionB ""
+    end
+  | _ -> OptionB ""
+[%%expect{|
```

Comment:
This one definitely seems fine.

Reply:
Right. Allowing that case used to break the associativity of `seq` but now that we make the matching with learned tags separate from `seq` this is fine. Allowing this now.